### ISSUE TITLE
[BUG] `CrossEntropyFunction.forward()` modifies input in-place without `ctx.mark_dirty()`, causing GPU memory leak

### ISSUE THREAD

**Issue by shen-jiabin** [NVIDIA/TransformerEngine#2899]
`CrossEntropyFunction` in `transformer_engine/pytorch/cross_entropy.py` modifies the input tensor **in-place** via Triton kernels (`online_softmax_kernel` + `cross_entropy_kernel`) but does **not** call `ctx.mark_dirty(inp)`. This violates the [PyTorch `torch.autograd.Function` contract](https://pytorch.org/docs/stable/autograd.html#torch.autograd.Function.forward) and causes `Function.apply()`'s C++ layer to permanently retain a reference to the input tensor, leaking GPU memory on every call.

## Root Cause

In `CrossEntropyFunction.forward()`:

```python
def forward(ctx, inp, target, ...):
    loss, inp = triton_cross_entropy.cross_entropy_forward(inp, target, ...)
    # ↑ Triton kernels overwrite inp data in-place (logits → softmax/gradient data)
    ctx.save_for_backward(inp.detach())
    # ↑ Missing: ctx.mark_dirty(inp)
    return loss
```

`triton_cross_entropy.cross_entropy_forward()` passes `inp` as `X_ptr` to `online_softmax_kernel` and `cross_entropy_kernel`, which overwrite the tensor's data in-place. However, since `ctx.mark_dirty(inp)` is never called, PyTorch's C++ `THPFunction_apply` retains a strong reference to the original input tensor for version-counter consistency checking. This reference is held by a C++ object and is invisible to Python's garbage collector.

## Impact

- **Memory leak**: Each call to `CrossEntropyFunction.apply()` leaks the full logits tensor (shape `[seq_len, vocab_size/tp]`, bf16). In our case, ~200-600 MiB per micro-batch, accumulating to 3-5 GiB per training step with 16 micro-batches.
- **OOM**: Training crashes after a few steps.
- **Affected configurations**: Any training using TE's cross entropy via `parallel_cross_entropy()` or `CrossEntropyFunction.apply()`, especially noticeable with gradient accumulation (multiple micro-batches per step).

## Reproduction

```python
import torch
import transformer_engine.pytorch.cross_entropy as te_ce

# Simulate repeated calls (as in gradient accumulation)
for i in range(16):
    logits = torch.randn(1024, 32000, dtype=torch.bfloat16, device='cuda', requires_grad=True)
    target = torch.randint(0, 32000, (1024,), device='cuda')
    loss = te_ce.parallel_cross_entropy(logits, target)
    loss.sum().backward()
    del loss, logits, target
    
    # Memory keeps growing despite del + backward
    print(f"Step {i}: {torch.cuda.memory_allocated()/1024**3:.2f} GiB")
```

## Environment

- TransformerEngine version: installed via pip (with PyTorch 2.10.0+cu128)
- PyTorch version: 2.10.0+cu128
- GPU: A800-80GB

## Workaround

Track the input tensor reference in a global deque during `forward()`, then `storage().resize_(0)` after backward completes:

```python
import collections
_leaked_ce_inputs = collections.deque()

def _drain_leaked_inputs():
    while _leaked_ce_inputs:
        t = _leaked_ce_inputs.popleft()
        if t.untyped_storage().size() > 0:
            t.storage().resize_(0)
        del t

# In CrossEntropyFunction.forward():
#   _leaked_ce_inputs.append(inp)  # before triton call

# In training loop, after backward_step():
#   _drain_leaked_inputs()
```

## Suggested Fix

Add `ctx.mark_dirty(inp)` in `CrossEntropyFunction.forward()` to properly inform PyTorch's autograd that the input was modified in-place. Note that `mark_dirty` requires the modified tensor to also be returned as an output, so the function signature may need adjustment. Alternatively, clone the input before passing to Triton kernels so the original input is not modified.

**Comment by shen-jiabin** [comment]
## Additional Environment Details

  | Component | Version |
  |-----------|---------|
  | TransformerEngine | 2.13.0 |
  | PyTorch | 2.10.0+cu128 |
  | Megatron Core | 0.18.0 |
  | CUDA | 12.8 |

  ## Related Issue: TE `OperationFuser._do_not_clear` interaction with external tensors

  While investigating this `CrossEntropyFunction` leak, we discovered a **second, separate memory leak** that involves TE's `OperationFuser`.

  ### Context

  When using `context_parallel_size > 1` in Megatron, `hidden_states` becomes a view (from sequence splitting across CP ranks). Megatron's
  `make_viewless_tensor()` (in `megatron/core/utils.py`) clones it to break the view relationship, producing a tensor with independent storage. This cloned
  tensor then flows into TE modules (LayerNorm, Attention, etc.) as input.

  ### TE's role in the leak

  In `OperationFuser.__call__()` → `_OperationFuserAutogradFunction.forward()`
  ([fuser.py:99-100](https://github.com/NVIDIA/TransformerEngine/blob/main/transformer_engine/pytorch/ops/fuser.py#L99-L100)):

  ```python
  # Mark input tensors as not deletable in backward
  for tensor in (input_,) + params_and_extra_inputs:
      tensor._do_not_clear = True
  ```

  This `_do_not_clear = True` attribute is checked by `clear_tensor_data()` in
  [utils.py:48](https://github.com/NVIDIA/TransformerEngine/blob/main/transformer_engine/pytorch/utils.py#L48):

  ```python
  if hasattr(t, "_do_not_clear"):
      continue  # skip clearing this tensor
  ```

  The `_do_not_clear` flag is **never removed** after backward completes. Combined with PyTorch's C++ autograd engine retaining references to the input
  tensor after backward, this means:

  1. C++ autograd holds a reference to the cloned `hidden_states` tensor
  2. `_do_not_clear` prevents any cleanup path from releasing its storage
  3. The tensor's GPU memory is leaked permanently (~30-1200 MiB per micro-batch depending on sequence length)

  ### Impact

  - **Leak rate**: ~0.5 GiB/step (with 16 micro-batches), leading to OOM within ~40 steps
  - **Affected configurations**: Any training with `context_parallel_size > 1` where `hidden_states` is a view, using TE modules downstream

  ### Suggestion

  Consider clearing `_do_not_clear` on input/output tensors at the end of `_OperationFuserAutogradFunction.backward()`, e.g.:

  ```python
  @staticmethod
  @torch.autograd.function.once_differentiable
  def backward(func_ctx, grad_output, *grad_extra_outputs):
      # ... existing backward logic ...

      # Clean up _do_not_clear flags (added at forward time)
      # to allow external cleanup paths to release tensor storage
      for ctx in basic_op_ctxs:
          if hasattr(ctx, '_input_tensors'):
              for t in ctx._input_tensors:
                  if hasattr(t, '_do_not_clear'):
                      delattr(t, '_do_not_clear')

      return (dx, None, None, *grad_params_flat, *grad_extra_inputs_flat)
  ```

  We plan to file a separate issue on Megatron-LM for the `make_viewless_tensor` side of this problem, but TE's `_do_not_clear` mechanism is a contributing 
  factor.
                                                                                                                                                            
  ---

**Comment by beccohov** [comment]
Hey,
I can work on this!
cc @sanandaraj5597 @ksivaman
